src/core/blueprint/wrapping_type.rs

```diff
@@ -173,3 +175,235 @@ impl From<String> for Type {
         Self::Named { name: value, non_null: false }
     }
 }
+
+impl Type {
+    fn merge(self, other: Self, non_null_merge: fn(bool, bool) -> bool) -> Valid<Self, String> {
+        match (self, other) {
+            (
+                Type::Named { name, non_null },
+                Type::Named { name: other_name, non_null: other_non_null },
+            ) => {
+                if name != other_name {
+                    return Valid::fail(format!(
+                        "Type mismatch: expected `{}`, got `{}`",
+                        &name, other_name
+                    ));
+                }
+
+                Valid::succeed(Type::Named {
+                    name,
+                    // non_null only if type is non_null for both sources
+                    non_null: non_null_merge(non_null, other_non_null),
+                })
+            }
+            (
+                Type::List { of_type, non_null },
+                Type::List { of_type: other_of_type, non_null: other_non_null },
+            ) => (*of_type)
+                .merge(*other_of_type, non_null_merge)
+                .map(|of_type| Type::List {
+                    of_type: Box::new(of_type),
+                    non_null: non_null_merge(non_null, other_non_null),
+                }),
+            _ => Valid::fail("Type mismatch: expected list, got singular value".to_string()),
+        }
+    }
+}
+
+impl FederatedMergeIntersection for Type {
+    /// Executes merge the way that the result type is non_null
+    /// if it is specified as non_null in at least one of the definitions.
+    /// That's a narrows merge i.e. the result narrows the input definitions
+    /// the way it could be handled by both self and other sources
+    fn federated_merge_intersection(self, other: Self) -> Valid<Self, String> {
+        #[inline]
+        fn non_null_merge(non_null: bool, other_non_null: bool) -> bool {
+            non_null || other_non_null
+        }
+
+        self.merge(other, non_null_merge)
+    }
+}
+
+impl FederatedMergeUnion for Type {
+    /// Executes merge the way that the result type is non_null only
+    /// if it is specified as non_null in both sources.
+    /// That's a wide merge i.e. the result wides the input definitions
+    /// the way it could be handled by both self and other sources
+    fn federated_merge_union(self, other: Self) -> Valid<Self, String> {
+        #[inline]
+        fn non_null_merge(non_null: bool, other_non_null: bool) -> bool {
+            non_null && other_non_null
+        }
+
+        self.merge(other, non_null_merge)
+    }
+}
```

Comment:
Optional: Can move the trait implementations for Type to `merge.rs` where all are defined.

Reply:
moved as suggested